### ISSUE TITLE
Assertj-Guava seems to have dependency to Guava Snapshot builds

### ISSUE THREAD

**Issue by geld0r** [joel-costigliola/assertj-guava#33]
I've just had the situation where updating dependencies in my Gradle project caused a new Snapshot build of Guava to be pulled in, even though I don't have that explicitly configured. In my case this caused an incompatibility with another 3rd party library.

After executing `gradle dependencies` I've gotten the following:
```
+--- org.assertj:assertj-guava:3.0.0
|    +--- com.google.guava:guava:[15.0,) -> 20.0-SNAPSHOT
|    \--- org.assertj:assertj-core:[3.0.0, 3.99.0] -> 3.5.2

```
It looks like assertj-guava pulls in newly released guava snapshot releases. Actually, it should pull any new version at all.
I am not very familiar with Maven and couldn't pinpoint the cause for this directly. Maybe this is caused in some upstream project?

Workaround for gradle:
```
compile "com.google.guava:guava:18.0"
testCompile ("org.assertj:assertj-guava:3.0.0")
{
    // Exclude guava in order to avoid newer versions to be pulled
    exclude module: 'guava'
}
```

**Comment by PascalSchumacher** [comment]
Hi geld0r,

assertj-guava defines range an open range guava versions it can work with starting from 15.0 (`[15.0,) `) 

If you use guava directly in your project, you should define the guava version your project requires as a direct dependency (according to maven/gradle "best practices"). That's what you did in line one of your workaround.

If you don't do this and the guava dependency of assertj-guava is the guava dependency closest to the root (your project), maven will use the newest guava version matching the version rage. That should be guava-19.0 at the moment. Normally snapshot dependencies are not used, because the are not in the maven central repository. So you or some dependency of your project add a snapshot repository.

To summarize this is not really a problem of asssertj-guava.

You should be able to remove the exclusion part of your workaround.

I hope this helps.

Cheers,
Pascal

**Comment by geld0r** [comment]
Hi Pascal,

thanks for the quick response!
To clarify, the `compile "com.google.guava:guava:18.0"` was already part of my build.gradle before I introduced the workaround.

I understand that assertj-guava doesn't define the new version itself and it is possible that one of my other dependencies does this.
What I don't understand however why my problem is solved with the above mentioned workaround then. I just verified that removing this brings back `guava:20.0-SNAPSHOT`.
From the output of `gradle dependencies` I don't see any other library that defines this.

If this is not reproducable for you or others, feel free to close this as invalid.

**Comment by PascalSchumacher** [comment]
Hi geld0r,

thanks for clarifying. I just created two simple projects.

With maven everything works as expected:

```xml
    <dependencies>
        <dependency>
            <groupId>org.assertj</groupId>
            <artifactId>assertj-guava</artifactId>
            <version>3.0.0</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>com.google.guava</groupId>
            <artifactId>guava</artifactId>
            <version>18.0</version>
        </dependency>
    </dependencies>
```
results in

```
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ assertj-guava-version-range-test ---
[INFO] com.test:assertj-guava-version-range-test:jar:0.0.1-SNAPSHOT
[INFO] +- org.assertj:assertj-guava:jar:3.0.0:test
[INFO] |  \- org.assertj:assertj-core:jar:3.5.2:test (version selected from constraint [3.0.0,3.99.0])
[INFO] \- com.google.guava:guava:jar:18.0:compile
```
The classpath is correct, as the declared guava version 18.0 is used.

But gradle handles this differently, so
```
dependencies {
    compile "com.google.guava:guava:18.0"
    testCompile "org.assertj:assertj-guava:3.0.0"
}
```
results in
```
compile - Dependencies for source set 'main'.
\--- com.google.guava:guava:18.0

compileClasspath - Compile classpath for source set 'main'.
\--- com.google.guava:guava:18.0

compileOnly - Compile dependencies for source set 'main'.
\--- com.google.guava:guava:18.0

default - Configuration for default artifacts.
\--- com.google.guava:guava:18.0

runtime - Runtime dependencies for source set 'main'.
\--- com.google.guava:guava:18.0

testCompile - Dependencies for source set 'test'.
+--- com.google.guava:guava:18.0 -> 19.0
\--- org.assertj:assertj-guava:3.0.0
     +--- com.google.guava:guava:[15.0,) -> 19.0
     \--- org.assertj:assertj-core:[3.0.0, 3.99.0] -> 3.5.2

testCompileClasspath - Compile classpath for source set 'test'.
+--- com.google.guava:guava:18.0 -> 19.0
\--- org.assertj:assertj-guava:3.0.0
     +--- com.google.guava:guava:[15.0,) -> 19.0
     \--- org.assertj:assertj-core:[3.0.0, 3.99.0] -> 3.5.2

testCompileOnly - Compile dependencies for source set 'test'.
+--- com.google.guava:guava:18.0 -> 19.0
\--- org.assertj:assertj-guava:3.0.0
     +--- com.google.guava:guava:[15.0,) -> 19.0
     \--- org.assertj:assertj-core:[3.0.0, 3.99.0] -> 3.5.2

testRuntime - Runtime dependencies for source set 'test'.
+--- com.google.guava:guava:18.0 -> 19.0
\--- org.assertj:assertj-guava:3.0.0
     +--- com.google.guava:guava:[15.0,) -> 19.0
     \--- org.assertj:assertj-core:[3.0.0, 3.99.0] -> 3.5.2
```
which means the test classpaths are wrong, as the newest guava-version is used (tested with 2.12 and 3.1) . This looks like gradle bug to me, maybe it is https://issues.gradle.org/browse/GRADLE-3392

**Comment by joel-costigliola** [comment]
@PascalSchumacher thanks for the investigation

**Comment by stevenschlansker** [comment]
This also causes problems with dependency version checkers such as https://github.com/ning/maven-dependency-versions-check-plugin :
```
[ERROR] Found a problem with the direct dependency com.google.guava:guava of the current project
  Expected version is 19.0
  Resolved version is 19.0
  Version 18.0 was expected by artifact: com.google.auto.service:auto-service
  Version 19.0 was expected by artifacts: org.jdbi:jdbi3, org.jdbi:jdbi3:tests
  Version 20.0-rc1 was expected by artifact: org.assertj:assertj-guava
```
even when you declare your own `<dependencyManagement>` for Guava.

I think this version range is more harmful than helpful.  Additional justification: https://ceylon-lang.org/blog/2014/12/01/version-ranges/

**Comment by joel-costigliola** [comment]
It seems that assertj-guava could indeed set a specific version of guava and let project override it with whatever version they want. I will use the mentioned maven plugin and do more testing but if things work I'm ok to remove the version range (I thought a version range would give flexibility but it appears I was mistaken).

**Comment by stevenschlansker** [comment]
I proposed #34 as a fix, let us know how testing goes.  Thanks! :)

**Comment by joel-costigliola** [comment]
Should be fixed as #34 has been merged.

**Comment by joel-costigliola** [comment]
@stevenschlansker @geld0r FYI, I have released 3.1.0 version with this fix yesterday.

**Comment by stevenschlansker** [comment]
Thanks!
